### PR TITLE
Add VA Auth0 integration

### PR DESCRIPTION
**Pull request by vibhanshub**

### CODE REVIEW

File: integrations/va-auth0/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                            }}
+                        />
+                    }
+                />
+                {!element.state.client_id ||
+                !element.state.client_secret ||
+                !element.state.issuer_base_url ? (
+                    <hint>
+                        <text style="bold">Enter values for the fields above and hit Save</text>
+                    </hint>
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:116]: I feel we need to stop being overly descriptive in our UI screens

File: integrations/va-auth0/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                ) : null}
+                <divider size="medium" />
+                <text>Enter the following URL as an allowed callback URL in Auth0:</text>
+                <text>{VACallbackURL}</text>
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:120]: This will not look great. Why not use  a `codeblock` that also allows the user to copy the whole thing easily?

File: integrations/va-auth0/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            if (context.environment.spaceInstallation?.space) {
+                const space = await context.api.spaces.getSpaceById(
+                    context.environment.spaceInstallation?.space
+                );
+                const spaceData = space.data;
+                const privateKey = context.environment.signingSecret;
+                let token;
+                try {
+                    token = await sign(
+                        { exp: Math.floor(Date.now() / 1000) + 2 * (60 * 60) },
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:144]: Why 2 hours? Imo we should avoid large expiry dates

File: integrations/va-okta/gitbook-manifest.yaml
Comment (taranvohra) [inline integrations/va-okta/gitbook-manifest.yaml:1]: Again separate PR

File: packages/runtime/src/integrations.ts [hunk 4]
```diff
@@ -108,6 +118,21 @@ export function createIntegration<Context extends RuntimeContext = RuntimeContex
...
+
+                logger.debug(
+                    `response ${resp.status} ${resp.statusText} Content-Type: ${resp.headers.get(
+                        'content-type'
+                    )}`
+                );
+
+                return resp;
+            }
+
```
Comment (taranvohra) [inline packages/runtime/src/integrations.ts:135]: This should be done in a separate PR with a changeset for `runtime` package so it can be published separately

File: spec/openapi.yaml
Comment (taranvohra) [inline spec/openapi.yaml:1]: Why? we should not commit this

File: integrations/va-auth0/gitbook-manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+name: VA-Auth0
+title: VA Auth0
+description: Visitor Authentication with Auth0
+externalLinks:
+    - label: Documentation
+      url: https://www.gitbook.com/integrations/heap
```
[the remark this answers is not in the document]
Comment (vibhanshub) [reply integrations/va-auth0/gitbook-manifest.yaml:6]: Removed

File: integrations/va-auth0/gitbook-manifest.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+visibility: public #unlisted
+script: ./src/index.tsx
+scopes:
+    - space:metadata:read
+    - space:visitor:auth
+    - space:content:read
+organization: w45sBUBc1JWE1ktYkQfI #d8f63b60-89ae-11e7-8574-5927d48c4877
```
[the remark this answers is not in the document]
Comment (vibhanshub) [reply integrations/va-auth0/gitbook-manifest.yaml:13]: This is the GitBook organization on my local environment. I will update to the gitbook org in prod with visibility set to public when the code changes have been reviewed.

I have this now because I need it to test the integrations locally (according to our development docs)

File: integrations/va-auth0/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                                status: 401,
+                            });
+                        });
+
+                    if ('access_token' in resp) {
+                        let url;
+                        if (request.query.state) {
+                            url = `${spaceData.urls?.published}${request.query.state}/?jwt_token=${token}`;
+                        } else {
+                            url = `${spaceData.urls?.published}/?jwt_token=${token}`;
```
[the remark this answers is not in the document]
Comment (vibhanshub) [reply integrations/va-auth0/src/index.tsx:211]: `request.query.state` contains the path the user wanted to go to before the auth flow started. We pass this path to the auth provider when the auth flow starts and the auth provider returns us this path when the auth flow completes.

So if a user tries to access a link behind VA, say: `docs.gitbook.com/my/docs`, we want the user to end up on the same link after authentication instead of redirecting them to `docs.gitbook.com` after authentication. 

`request.query.state` would be `my/docs` in this case.

**Comment by vibhanshub** [comment]
@taranvohra here are the separated PRs:
Updating Runtime to include fetch_visitor_authentication event with changeset: https://github.com/GitbookIO/integrations/pull/369

Okta integration: https://github.com/GitbookIO/integrations/pull/368

Azure integration: https://github.com/GitbookIO/integrations/pull/367

File: integrations/va-auth0/gitbook-manifest.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    Visitor Authentication allows you to publish content behind an authentication wall so that your content is only accessible to selected people.
+
+    This integration lets you control access to your published content as determined by Auth0.
+
+    # Configure
+    Install this integration on a space and then populate the configuration screen with the details of your Auth0 application and Auth0 instance
+    Then publish the space with Visitor Authentication, choose this integration as the backend, and hit save.
+
+    Your space is now published with Visitor Authentication using Auth0.
+
```
Comment (taranvohra) [inline integrations/va-auth0/gitbook-manifest.yaml:23]: I'd suggest we confirm with marketing once and get their opinion on the copy

File: integrations/va-auth0/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    },
+    render: async (element, context) => {
+        const VACallbackURL = `${context.environment.spaceInstallation?.urls?.publicEndpoint}/visitor-auth/response`;
+        return (
+            <block>
+                <input
+                    label="Enter Client ID"
+                    hint={
+                        <text>
+                            Enter the unique identifier of your Auth0 application.
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:82]:
```suggestion
                            The unique identifier of your Auth0 application.
```

File: integrations/va-auth0/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+                );
+                return element;
+        }
+    },
+    render: async (element, context) => {
+        const VACallbackURL = `${context.environment.spaceInstallation?.urls?.publicEndpoint}/visitor-auth/response`;
+        return (
+            <block>
+                <input
+                    label="Enter Client ID"
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:79]:
```suggestion
                    label="Client ID"
```

File: integrations/va-auth0/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+                                {' '}
+                                More Details
+                            </link>
+                        </text>
+                    }
+                    element={<textinput state="client_id" placeholder="Client ID" />}
+                />
+
+                <input
+                    label="Enter Your Auth0 Domain"
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:97]:
```suggestion
                    label="Auth0 Domain"
```

File: integrations/va-auth0/src/index.tsx [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                    hint={
+                        <text>
+                            Enter the Auth0 domain (also known as tenant).
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:100]:
```suggestion
                            The Auth0 domain (also known as tenant).
```

File: integrations/va-auth0/src/index.tsx [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+                                {' '}
+                                More Details
+                            </link>
+                        </text>
+                    }
+                    element={<textinput state="issuer_base_url" placeholder="Domain" />}
+                />
+
+                <input
+                    label="Enter Client Secret"
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:115]:
```suggestion
                    label="Client Secret"
```

File: integrations/va-auth0/src/index.tsx [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+                    hint={
+                        <text>
+                            Enter the secret used for signing and validating tokens.
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:118]:
```suggestion
                            The secret used for signing and validating tokens.
```

File: integrations/va-auth0/src/index.tsx [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+                            label="Save"
+                            tooltip="Save configuration"
+                            onPress={{
+                                action: 'save.config',
+                            }}
+                        />
+                    }
+                />
+                <divider size="medium" />
+                <text>Enter the following URL as an allowed callback URL in Auth0:</text>
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:148]:
```suggestion
                <text>The following URL needs to be saved as an allowed callback URL in Auth0:</text>
```

File: integrations/va-auth0/src/index.tsx [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+                                status: 401,
+                            });
+                        });
+
+                    if ('access_token' in resp) {
+                        let url;
+                        if (request.query.state) {
+                            url = `${spaceData.urls?.published}${request.query.state}/?jwt_token=${token}`;
+                        } else {
+                            url = `${spaceData.urls?.published}/?jwt_token=${token}`;
```
Comment (taranvohra) [inline integrations/va-auth0/src/index.tsx:211]: not blocking but using the `URL` constructor is usually a bit more safe

File: package-lock.json
Comment (taranvohra) [inline package-lock.json:1]: This `package-lock` is incorrectly computed from last time. You need to run `npm install` again to make sure only `va-auth0` and it's deps are being added.

File: integrations/va-auth0/gitbook-manifest.yaml (hunk 8 above)
Comment (taranvohra) [reply integrations/va-auth0/gitbook-manifest.yaml:23]: We should also get `previewImages` from them imo

**Review by taranvohra — APPROVED** [review]

File: integrations/va-auth0/gitbook-manifest.yaml (hunk 8 above)
Comment (vibhanshub) [reply integrations/va-auth0/gitbook-manifest.yaml:23]: I've asked Marketing for their opinion and preview images, but I think we can add previewImages to the integration later and get this merged.

Comment (vibhanshub) [reply integrations/va-auth0/gitbook-manifest.yaml]: Marketing suggested changes to the copy, which I have now made.
